crates/turborepo-lib/src/runsummary/execution.rs

```diff
@@ -0,0 +1,279 @@
+#[derive(Debug)]
+pub struct ExecutionSummary {
+    // this thread handles the state management
+    state_thread: tokio::task::JoinHandle<SummaryState>,
+    sender: mpsc::Sender<Message>,
+}
+
+/// The final states of all task executions
+#[derive(Debug, Default, Clone, Copy)]
+pub struct SummaryState {
+    pub attempted: usize,
+    pub failed: usize,
+    pub cached: usize,
+    pub success: usize,
+}
+
+impl SummaryState {
+    fn handle_event(&mut self, event: Event) {
+        match event {
+            Event::Building => self.attempted += 1,
+            Event::BuildFailed => self.failed += 1,
+            Event::Cached => self.cached += 1,
+            Event::Built => self.success += 1,
+        }
+    }
+}
+
+/// A tracker constructed for each task and used to communicate task events back
+/// to the exeuction summary.
+pub struct Tracker<T> {
+    sender: mpsc::Sender<Message>,
+    started_at: T,
+    // task_id is only used as a name for the event in the chrometracing profile
+    #[allow(dead_code)]
+    task_id: TaskId<'static>,
+}
+
+#[derive(Debug, Clone, Copy)]
+enum Event {
+    Building,
+    BuildFailed,
+    Cached,
+    Built,
+}
+
+enum ExecutionState {
+    Canceled,
+    Built { exit_code: u32 },
+    Cached,
+    BuildFailed { exit_code: u32, err: String },
+}
+
+pub struct TaskExecutionSummary {
+    started_at: Instant,
+    ended_at: Instant,
+    state: ExecutionState,
+}
+
+impl TaskExecutionSummary {
+    pub fn exit_code(&self) -> Option<u32> {
+        match self.state {
+            ExecutionState::BuildFailed { exit_code, .. } | ExecutionState::Built { exit_code } => {
+                Some(exit_code)
+            }
+            _ => None,
+        }
+    }
+
+    pub fn duration(&self) -> Duration {
+        self.ended_at.duration_since(self.started_at)
+    }
+}
+
+impl ExecutionSummary {
+    pub fn new() -> Self {
+        // This buffer size is probably overkill, but since messages are only a byte
+        // it's worth the extra memory to avoid the channel filling up.
+        let (sender, mut receiver) = mpsc::channel(128);
+        let state_thread = tokio::spawn(async move {
+            let mut state = SummaryState::default();
+            while let Some(event) = receiver.recv().await {
+                state.handle_event(event);
+            }
+            state
+        });
+        Self {
+            state_thread,
+            sender,
+        }
+    }
+
+    // Produce a tracker for the task
+    pub fn tracker(&self, task_id: TaskId<'static>) -> Tracker<()> {
+        Tracker {
+            sender: self.sender.clone(),
+            task_id,
+            started_at: (),
+        }
+    }
+
+    pub async fn finish(self) -> Result<SummaryState, tokio::task::JoinError> {
+        let Self {
+            state_thread,
+            sender,
+        } = self;
+        // We drop the sender so the channel closes once all trackers have finished
+        // we don't explicitly close as that would
```

Comment:
that would?

Reply:
I like my comments to have a mystery about them